# T05 MOTOR TREND CAR ROAD TESTS
## Paola A. Figueroa Álvarez
## 16/sep/2025

### 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

Calcula el R2 e interpreta los signos de los betas.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from scipy import stats 
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import Ridge

In [2]:
data= pd.read_excel("/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/Motor Trend Car Road Tests.xlsx")
data = data.drop(columns="model")
data.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [3]:
data2=data.dropna().drop_duplicates()
data2

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0   21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1   21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2   22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3   21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4   18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2
5   18.1    6  225.0  105  2.76  3.460  20.22   1   0     3     1
6   14.3    8  360.0  245  3.21  3.570  15.84   0   0     3     4
7   24.4    4  146.7   62  3.69  3.190  20.00   1   0     4     2
8   22.8    4  140.8   95  3.92  3.150  22.90   1   0     4     2
9   19.2    6  167.6  123  3.92  3.440  18.30   1   0     4     4
10  17.8    6  167.6  123  3.92  3.440  18.90   1   0     4     4
11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3     3
12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3     3
13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3     3
14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3     4
15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3     4
16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3     4
17  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4     1
18  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4     2
19  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4     1
20  21.5    4  120.1   97  3.70  2.465  20.01   1   0     3     1
21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3     2
22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3     2
23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3     4
24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3     2
25  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4     1
26  26.0    4  120.3   91  4.43  2.140  16.70   0   1     5     2
27  30.4    4   95.1  113  3.77  1.513  16.90   1   1     5     2
28  15.8    8  351.0  264  4.22  3.170  14.50   0   1     5     4
29  19.7    6  145.0  175  3.62  2.770  15.50   0   1     5     6
30  15.0    8  301.0  335  3.54  3.570  14.60   0   1     5     8
31  21.4    4  121.0  109  4.11  2.780  18.60   1   1     4     2

In [4]:
cols=data2.columns
cols=cols.drop("mpg")
x_list = []

for col in cols:
    x_list.append(data2[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = data2["mpg"].values.reshape(-1, 1)

scaler=StandardScaler().fit(X)
Xescalada=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xescalada))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        19:35:01   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.468     42.884      0.000      19.116      21.065
x1            -0.1959      1.837     -0.107      0.916      -4.016       3.624
x2             1.6267      2.178      0.747      0.463      -2.903       6.157
x3            -1.4497      1.469     -0.987      0.335      -4.505       1.605
x4             0.4142      0.861      0.481      0.635      -1.376       2.204
x5            -3.5780      1.824     -1.961      0.063      -7.372       0.216
x6             1.4440      1.285      1.123      0.274      -1.229       4.117
x7             0.1576      1.044      0.151      0.881      -2.013       2.329
x8             1.2378      1.010      1.225      0.234      -0.863       3.338
x9             0.4760      1.084      0.439      0.665      -1.779       2.731
x10           -0.3170      1.318     -0.241      0.812      -3.057       2.423
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                         15.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [5]:
X_mpg = data2.drop(columns=['mpg']).values
y_mpg = data2['mpg'].values

In [6]:
X_mpg_train, X_mpg_test, y_mpg_train, y_mpg_test = train_test_split(X_mpg, y_mpg, train_size=0.4, random_state=137)
scaler = StandardScaler().fit(X_mpg_train)
X_mpg_train_scaled = scaler.transform(X_mpg_train)
X_mpg_test_scaled = scaler.transform(X_mpg_test)

In [7]:
X_mpg_train_sm = sm.add_constant(X_mpg_train_scaled)
modelo_mpg = sm.OLS(y_mpg_train, X_mpg_train_sm).fit()

In [8]:
modelo_mpg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Wed, 17 Sep 2025   Prob (F-statistic):              0.242
Time:                        19:35:01   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1            40.7571     54.010      0.755      0.588    -645.499     727.013
x2            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x3           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x4            14.1135     13.469      1.048      0.485    -157.022     185.249
x5            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x6             1.5727      3.475      0.453      0.729     -42.583      45.729
x7            32.4926     41.262      0.787      0.575    -491.795     556.780
x8            36.1394     42.669      0.847      0.553    -506.022     578.301
x9           -38.3551     43.625     -0.879      0.541    -592.668     515.958
x10           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                         382.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Desgloce de los coeficientes:
- x1 - cyl (40.7571): Por cada cilindro adicional, el mpg aumenta en 40.76 unidades. 
- x2 - disp (-2.4635): Por cada pulgada cúbica adicional de desplazamiento, el mpg disminuye en 2.46 unidades. Motores más grandes consumen más combustible.
- x3 - hp (-10.1581): Por cada fuerza adicional, el mpg disminuye en 10.16 unidades. A mayor potencia, mayor consumo.
- x4 - drat (14.1135): Por cada unidad adicional en la relación del eje trasero, el mpg aumenta en 14.11 unidades. Una mejor relación puede mejorar eficiencia.
- x5 - wt (-0.8005): Por cada peso adicional, el mpg disminuye en 0.80 unidades, ya que los autos más pesados pesados consumen más combustible.
- x6 - qsec (1.5727): Por cada segundo adicional en tiempo, el mpg aumenta en 1.57 unidades. Los autos más lentos suelen ser más eficientes en combustible (gastan menos)
- x7 - vs (32.4926): Un mejor motor aumenta el combustible.
- x8 - am (36.1394): una mejor transmisión aumenta el  mpg en 36.14 unidades.
- x9 - gear (-38.3551): Por cada velocidad adicional, el mpg disminuye en 38.36 unidades.
- x10 - carb (11.9707): Por cada carburador adicional, el mpg aumenta en 11.97 unidades.

Podemos ver que los p-values son muy altos (mayores al 0.05), lo que significa que es pura chiripa, es decir, ninguna variable es estadísticamente significativa, lo que sugiere que ninguna de las variables tiene efecto fuerte sobre Y.
El modelo logra explicar el 99% de la variabilidad de mpg.

In [9]:
# Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.
X_mpg_test_sm = sm.add_constant(X_mpg_test_scaled)
y_mpg_pred = modelo_mpg.predict(X_mpg_test_sm)
r2_mpg_train= modelo_mpg.rsquared
r2_mpg_test= r2_score(y_mpg_test, y_mpg_pred)
print(f"R2 de entrenamiento: {r2_mpg_train}")
print(f"R2 de prueba: {r2_mpg_test}")

R2 de entrenamiento: 0.9900805114414537
R2 de prueba: -54.071785204187904


In [10]:
# Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.
lambdas = [0.01, 0.1, 1, 10, 100]
resultados_mpg_train={}
resultados_mpg_test={}
for l in lambdas:
    modelo_ridge = Ridge(alpha=l, fit_intercept=True)
    modelo_ridge.fit(X_mpg_train_scaled, y_mpg_train)
    y_mpg_train_ridge = modelo_ridge.predict(X_mpg_train_scaled)
    y_mpg_test_ridge= modelo_ridge.predict(X_mpg_test_scaled)
    # calculo de r2
    r2_train_ridge = r2_score(y_mpg_train, y_mpg_train_ridge)
    r2_test_ridge = r2_score(y_mpg_test, y_mpg_test_ridge)
    resultados_mpg_train[l] = (r2_train_ridge)
    resultados_mpg_test[l]=(r2_test_ridge)
    print(f"Lambda: {l}, R2 Entrenamiento: {r2_train_ridge}, R2 Prueba: {r2_test_ridge}")
    

Lambda: 0.01, R2 Entrenamiento: 0.98418950094601, R2 Prueba: 0.021165506080776608
Lambda: 0.1, R2 Entrenamiento: 0.9805234796986708, R2 Prueba: 0.23411656981043205
Lambda: 1, R2 Entrenamiento: 0.9386746881405903, R2 Prueba: 0.5910911509384542
Lambda: 10, R2 Entrenamiento: 0.8637598356203311, R2 Prueba: 0.7632502786243675
Lambda: 100, R2 Entrenamiento: 0.5184051943511396, R2 Prueba: 0.3498417578247329


Dado que estamos buscando el modelo que mejor generalice, estamos buscando un r2 alto en la prueba porque refleja la capacidad de predecir sobre nuevos datos. Un R2 alto en entrenamiento solo es útil si también se mantiene alto en prueba, de lo contrario sería un overfit. Lo ideal es un R2 decente en entrenamiento y lo más alto posible en prueba, manteniendo ambos relativamente cercanos.Considero que la mejor penalización fue la de lambda= 1, ya que explica el 94%% del modelo de entrenamiento y el 60% del modelo de prueba. La de lambda 0.1 también es buena e incluso tiene una mayor explicación sobre el modelo de entrenamiento pero menor sobre la de prueba, lo cual no es muy confiable, considerando que puede haber overfitting. 

###  Repite el ejercicio anterior usando 'qsec' como salida.

In [11]:
cols=data2.columns
cols=cols.drop("qsec")
x_list = []

for col in cols:
    x_list.append(data2[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = data2["qsec"].values.reshape(-1, 1)

scaler=StandardScaler().fit(X)
Xescalada=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xescalada))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        19:35:01   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.136    131.375      0.000      17.566      18.131
x1             0.4096      0.365      1.123      0.274      -0.349       1.168
x2            -0.6375      0.514     -1.239      0.229      -1.707       0.432
x3            -0.9150      0.608     -1.505      0.147      -2.180       0.350
x4            -0.1054      0.435     -0.242      0.811      -1.010       0.800
x5            -0.0690      0.250     -0.275      0.786      -0.590       0.452
x6             1.4410      0.482      2.990      0.007       0.439       2.443
x7             0.4812      0.284      1.694      0.105      -0.110       1.072
x8            -0.4426      0.287     -1.540      0.139      -1.040       0.155
x9            -0.1462      0.314     -0.465      0.647      -0.800       0.507
x10           -0.4350      0.371     -1.174      0.254      -1.206       0.336
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                         15.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [12]:
X_qsec = data2.drop(columns=['qsec']).values
y_qsec = data2['qsec'].values
data2.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [13]:
X_qsec_train, X_qsec_test, y_qsec_train, y_qsec_test = train_test_split(X_qsec, y_qsec, train_size=0.4, random_state=137)
scaler = StandardScaler().fit(X_qsec_train)
X_qsec_train_scaled = scaler.transform(X_qsec_train)
X_qsec_test_scaled = scaler.transform(X_qsec_test)

In [14]:
X_qsec_train_sm = sm.add_constant(X_qsec_train_scaled)
modelo_qsec = sm.OLS(y_qsec_train, X_qsec_train_sm).fit()
modelo_qsec.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Wed, 17 Sep 2025   Prob (F-statistic):              0.349
Time:                        19:35:01   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.009      14.232      20.515
x1             1.0235      2.262      0.453      0.729     -27.713      29.760
x2            -0.3081     30.092     -0.010      0.993    -382.663     382.047
x3             0.2789      1.711      0.163      0.897     -21.465      22.023
x4            -1.0127      9.720     -0.104      0.934    -124.519     122.494
x5            -1.0356      8.615     -0.120      0.924    -110.497     108.426
x6             0.5976      1.295      0.461      0.725     -15.861      17.057
x7             0.2129     23.358      0.009      0.994    -296.578     297.004
x8            -0.3877     24.867     -0.016      0.990    -316.349     315.574
x9             0.6755     25.827      0.026      0.983    -327.485     328.836
x10           -0.1562      9.422     -0.017      0.989    -119.876     119.563
==============================================================================
Omnibus:                       12.359   Durbin-Watson:                   2.150
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.063
Skew:                           1.532   Prob(JB):                       0.0293
Kurtosis:                       5.178   Cond. No.                         517.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

 Interpretación de resultados
- x1 - mpg (1.0235): Por cada milla por galón adicional, el qsec aumenta en 1.02 segundos. Autos más eficientes en combustible tienden a ser más lentos (menos deportivos).
- x2 - cyl (-0.3081): Por cada cilindro adicional, el qsec disminuye en 0.31 segundos; Más cilindros hacen al auto más rápido.
- x3 - disp (0.2789): Por cada pulgada cúbica adicional de desplazamiento, el qsec aumenta en 0.28 segundos. 
- x4 - hp (-1.0127): Por cada fuerza adicional, el qsec disminuye en 1.01 segundos; Más potencia hace al auto más rápido.
- x5 - drat (-1.0356): Por cada unidad adicional en la relación del eje trasero, el qsec disminuye en 1.04 segundos. Una mejor relación de transmisión mejora la velocidad.
- x6 - wt (0.5976): Por cada peso adicional, el qsec aumenta en 0.60 segundos. Los autos más pesados son más lentos.
- x7 - vs (0.2129):Un mejor motor mejora el tiempo
- x8 - am (-0.3877): Transmisión manual vs automática disminuye el qsec en 0.39 segundos. Una transmisión manual suele ser más rápida.
- x9 - gear (0.6755): Por cada velocidad adicional, el qsec aumenta en 0.68 segundos. 
- x10 - carb (-0.1562): Por cada carburador adicional, el qsec disminuye en 0.16 segundos. Más carburadores dan más potencia y, por lo tanto, más velocidad.

Usando este modelo con la variable constante de qsec, tenemos una r2 de 97%, lo cual es menor a la anterior. Viendo los p-values, todos son, al igual que en el modelo anterior, muy altos, lo que suguiere que todos son de pura chiripa; No podemos estar seguros de que alguna variable tenga un efecto real sobre qsec


In [15]:
# Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.
X_qsec_test_sm = sm.add_constant(X_qsec_test_scaled)
y_qsec_pred = modelo_qsec.predict(X_qsec_test_sm)
r2_qsec_train= modelo_qsec.rsquared
r2_qsec_test= r2_score(y_qsec_test, y_qsec_pred)
print(f"Error de entrenamiento: {r2_qsec_train}")
print(f"Error de prueba: {r2_qsec_test}")

Error de entrenamiento: 0.9787621925230521
Error de prueba: 0.3761038962001224


In [16]:
# Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.
lambdasq = [0.01, 0.1, 1, 10, 100]
resultados_qsec_train={}
resultados_qsec_test={}
for l in lambdasq:
    modelo_ridge_qsec = Ridge(alpha=l, fit_intercept=True)
    modelo_ridge_qsec.fit(X_qsec_train_scaled, y_qsec_train)
    y_qsec_train_ridge = modelo_ridge_qsec.predict(X_qsec_train_scaled)
    y_qsec_test_ridge= modelo_ridge_qsec.predict(X_qsec_test_scaled)
    # calculo de r2
    r2_train_ridge_qsec = r2_score(y_qsec_train, y_qsec_train_ridge)
    r2_test_ridge_qsec = r2_score(y_qsec_test, y_qsec_test_ridge)
    resultados_qsec_train[l] = (r2_train_ridge_qsec)
    resultados_qsec_test[l]=(r2_test_ridge_qsec)
    print(f"Lambda: {l}, R2 Entrenamiento: {r2_train_ridge_qsec}, R2 Prueba: {r2_test_ridge_qsec}")

Lambda: 0.01, R2 Entrenamiento: 0.978691352833023, R2 Prueba: 0.46413423269630893
Lambda: 0.1, R2 Entrenamiento: 0.9773047022861615, R2 Prueba: 0.5897331179193346
Lambda: 1, R2 Entrenamiento: 0.9671802657602482, R2 Prueba: 0.6535717344645446
Lambda: 10, R2 Entrenamiento: 0.8843606593441767, R2 Prueba: 0.6093368477249674
Lambda: 100, R2 Entrenamiento: 0.445390477416979, R2 Prueba: 0.24011778536032258


la mejor penalización fue la de lambda= 0.1, ya que explica el 97% del modelo de entrenamiento y el 59% del modelo de prueba. La de lambda 1 también es buena e incluso tiene una menor explicación sobre el modelo de entrenamiento pero mayor sobre la de prueba. A comparación del modelo sin penalizar, la r2 del entrenamiento es un poquito mejor, pero su r2 de prueba es mucho menor, lo que significa que al penalizar, los nuevos datos se pueden predecir/explicar mejor.

### 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

In [17]:
data_mpg_dummies = pd.get_dummies(data2, columns = ['cyl', 'gear', 'carb'])
data_mpg_dummies.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1  21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2  22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3  21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4  18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   

   gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0    True   False   False   False   False    True   False   False  
1    True   False   False   False   False    True   False   False  
2    True   False    True   False   False   False   False   False  
3   False   False    True   False   False   False   False   False  
4   False   False   False    True   False   False   False   False

In [18]:
cols=data_mpg_dummies.columns
cols=cols.drop("mpg")
x_list = []

for col in cols:
    x_list.append(data_mpg_dummies[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = data_mpg_dummies["mpg"].values.reshape(-1, 1)

scaler=StandardScaler().fit(X)
Xescalada=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xescalada))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()             

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           0.000124
Time:                        19:35:01   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.501     40.114      0.000      19.023      21.158
x1             4.3362      3.891      1.114      0.283      -3.958      12.630
x2            -4.7580      2.661     -1.788      0.094     -10.429       0.913
x3             0.6225      1.307      0.476      0.641      -2.163       3.408
x4            -4.3624      2.445     -1.784      0.095      -9.574       0.849
x5             0.6470      1.645      0.393      0.700      -2.860       4.154
x6             0.9579      1.424      0.672      0.512      -2.078       3.994
x7             0.5953      1.578      0.377      0.711      -2.769       3.959
x8             0.3957      1.593      0.248      0.807      -3.001       3.792
x9            -0.7505      0.826     -0.909      0.378      -2.511       1.010
x10            0.2466      1.934      0.127      0.900      -3.876       4.369
x11           -0.4822      1.028     -0.469      0.646      -2.673       1.708
x12            0.0717      0.935      0.077      0.940      -1.920       2.064
x13            0.5672      0.827      0.685      0.503      -1.196       2.331
x14           -0.3513      1.056     -0.333      0.744      -2.602       1.899
x15           -0.8478      0.803     -1.055      0.308      -2.560       0.864
x16            0.6267      0.870      0.721      0.482      -1.227       2.480
x17            0.1120      1.009      0.111      0.913      -2.039       2.263
x18            0.6312      0.803      0.786      0.444      -1.081       2.344
x19            1.1137      1.130      0.986      0.340      -1.295       3.522
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     4.88e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is  1e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [19]:
X_mpg_dummies = data_mpg_dummies.drop(columns=['mpg']).values
y_mpg_dummies = data_mpg_dummies['mpg'].values

In [20]:
X_mpg_dummies_train, X_mpg_dummies_test, y_mpg_dummies_train, y_mpg_dummies_test = train_test_split(X_mpg_dummies, y_mpg_dummies, train_size=0.4, random_state=137)
scaler = StandardScaler().fit(X_mpg_dummies_train)
X_mpg_dummies_train_scaled = scaler.transform(X_mpg_dummies_train)
X_mpg_dummies_test_scaled = scaler.transform(X_mpg_dummies_test)

In [21]:
X_mpg_dummies_train_sm = sm.add_constant(X_mpg_dummies_train_scaled)
modelo_mpg_dummies = sm.OLS(y_mpg_dummies_train, X_mpg_dummies_train_sm).fit()
data_mpg_dummies

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0   21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1   21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2   22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3   21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4   18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   
5   18.1  225.0  105  2.76  3.460  20.22   1   0  False   True  False    True   
6   14.3  360.0  245  3.21  3.570  15.84   0   0  False  False   True    True   
7   24.4  146.7   62  3.69  3.190  20.00   1   0   True  False  False   False   
8   22.8  140.8   95  3.92  3.150  22.90   1   0   True  False  False   False   
9   19.2  167.6  123  3.92  3.440  18.30   1   0  False   True  False   False   
10  17.8  167.6  123  3.92  3.440  18.90   1   0  False   True  False   False   
11  16.4  275.8  180  3.07  4.070  17.40   0   0  False  False   True    True   
12  17.3  275.8  180  3.07  3.730  17.60   0   0  False  False   True    True   
13  15.2  275.8  180  3.07  3.780  18.00   0   0  False  False   True    True   
14  10.4  472.0  205  2.93  5.250  17.98   0   0  False  False   True    True   
15  10.4  460.0  215  3.00  5.424  17.82   0   0  False  False   True    True   
16  14.7  440.0  230  3.23  5.345  17.42   0   0  False  False   True    True   
17  32.4   78.7   66  4.08  2.200  19.47   1   1   True  False  False   False   
18  30.4   75.7   52  4.93  1.615  18.52   1   1   True  False  False   False   
19  33.9   71.1   65  4.22  1.835  19.90   1   1   True  False  False   False   
20  21.5  120.1   97  3.70  2.465  20.01   1   0   True  False  False    True   
21  15.5  318.0  150  2.76  3.520  16.87   0   0  False  False   True    True   
22  15.2  304.0  150  3.15  3.435  17.30   0   0  False  False   True    True   
23  13.3  350.0  245  3.73  3.840  15.41   0   0  False  False   True    True   
24  19.2  400.0  175  3.08  3.845  17.05   0   0  False  False   True    True   
25  27.3   79.0   66  4.08  1.935  18.90   1   1   True  False  False   False   
26  26.0  120.3   91  4.43  2.140  16.70   0   1   True  False  False   False   
27  30.4   95.1  113  3.77  1.513  16.90   1   1   True  False  False   False   
28  15.8  351.0  264  4.22  3.170  14.50   0   1  False  False   True   False   
29  19.7  145.0  175  3.62  2.770  15.50   0   1  False   True  False   False   
30  15.0  301.0  335  3.54  3.570  14.60   0   1  False  False   True   False   
31  21.4  121.0  109  4.11  2.780  18.60   1   1   True  False  False   False   

    gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0     True   False   False   False   False    True   False   False  
1     True   False   False   False   False    True   False   False  
2     True   False    True   False   False   False   False   False  
3    False   False    True   False   False   False   False   False  
4    False   False   False    True   False   False   False   False  
5    False   False    True   False   False   False   False   False  
6    False   False   False   False   False    True   False   False  
7     True   False   False    True   False   False   False   False  
8     True   False   False    True   False   False   False   False  
9     True   False   False   False   False    True   False   False  
10    True   False   False   False   False    True   False   False  
11   False   False   False   False    True   False   False   False  
12   False   False   False   False    True   False   False   False  
13   False   False   False   False    True   False   False   False  
14   False   False   False   False   False    True   False   False  
15   False   False   False   False   False    True   False   False  
16   False   False   False   False   False    True   False   False  
17    True   False    True   False   False   False   False   False  
18    True   False  

### Calcula el R2 e interpreta los signos de los betas.

In [22]:
modelo_mpg_dummies.summary()

/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = sel

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Wed, 17 Sep 2025   Prob (F-statistic):                nan
Time:                        19:35:01   Log-Likelihood:                 364.24
No. Observations:                  12   AIC:                            -704.5
Df Residuals:                       0   BIC:                            -698.7
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333        inf          0        nan         nan         nan
x1            -2.8807        inf         -0        nan         nan         nan
x2             1.1616        inf          0        nan         nan         nan
x3             3.8237        inf          0        nan         nan         nan
x4             2.0417        inf          0        nan         nan         nan
x5            -1.5490        inf         -0        nan         nan         nan
x6            -0.3475        inf         -0        nan         nan         nan
x7             0.5954        inf          0        nan         nan         nan
x8             1.6043        inf          0        nan         nan         nan
x9             0.0494        inf          0        nan         nan         nan
x10           -0.9860        inf         -0        nan         nan         nan
x11            1.4522        inf          0        nan         nan         nan
x12            0.1433        inf          0        nan         nan         nan
x13           -2.0876        inf         -0        nan         nan         nan
x14            2.2705        inf          0        nan         nan         nan
x15            0.0735        inf          0        nan         nan         nan
x16           -0.1647        inf         -0        nan         nan         nan
x17           -1.3885        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19           -0.4010        inf         -0        nan         nan         nan
==============================================================================
Omnibus:                        1.144   Durbin-Watson:                   1.521
Prob(Omnibus):                  0.564   Jarque-Bera (JB):                0.833
Skew:                          -0.353   Prob(JB):                        0.660
Kurtosis:                       1.920   Cond. No.                         36.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

El modelo presenta un ajuste perfecto, lo cual es raro. La ausencia de p-values significa que el modelo está "roto" - hay demasiadas variables para tan pocos datos, con posible overfit, ya que el modelo es demasiado bueno para ajusarse al comportamiento real. Los coeficientes que aparecen no se pueden interpretar de manera confiable porque el modelo no puede determinar cuál variable realmente está causando qué efecto - todas están mezcladas.

### 2.1.2 Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [23]:
X_mpg_dummies_test_sm = sm.add_constant(X_mpg_dummies_test_scaled, has_constant='add')
y_mpg_dummies_pred = modelo_mpg_dummies.predict(X_mpg_dummies_test_sm)

In [24]:
r2_mpg_dummies_train = modelo_mpg_dummies.rsquared
r2_mpg_dummies_test = r2_score(y_mpg_dummies_test, y_mpg_dummies_pred)
print(f'R2 entrenamiento: {r2_mpg_dummies_train}')
print(f'R2 prueba: {r2_mpg_dummies_test}')

R2 entrenamiento: 1.0
R2 prueba: 0.12691935910997298


Como ya habíamos analizado, una r2 alta en entretenimiento pero mala en la prueba indica un overfitting; El modelo está capturando demasiado el ruido de los datos de entrenamiento y no logra predecir bien nuevos datos.

### 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

In [25]:
data_qsec_dummies = pd.get_dummies(data2, columns = ['cyl', 'gear', 'carb'])
data_qsec_dummies.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1  21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2  22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3  21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4  18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   

   gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0    True   False   False   False   False    True   False   False  
1    True   False   False   False   False    True   False   False  
2    True   False    True   False   False   False   False   False  
3   False   False    True   False   False   False   False   False  
4   False   False   False    True   False   False   False   False

In [26]:
cols=data_qsec_dummies.columns
cols=cols.drop("qsec")
x_list = []

for col in cols:
    x_list.append(data_qsec_dummies[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = data_qsec_dummies["qsec"].values.reshape(-1, 1)

scaler=StandardScaler().fit(X)
Xescalada=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xescalada))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()   

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        19:35:01   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.138    129.771      0.000      17.556      18.142
x1             0.1646      0.418      0.393      0.700      -0.727       1.057
x2             0.4308      1.106      0.389      0.703      -1.927       2.789
x3            -0.1394      0.804     -0.173      0.865      -1.853       1.574
x4             0.0568      0.361      0.157      0.877      -0.713       0.827
x5             0.7805      0.711      1.097      0.290      -0.735       2.297
x6             0.1318      0.396      0.333      0.744      -0.711       0.975
x7            -0.8321      0.379     -2.197      0.044      -1.639      -0.025
x8             0.6792      0.402      1.690      0.112      -0.177       1.536
x9             0.1346      0.230      0.584      0.568      -0.357       0.626
x10           -0.7624      0.494     -1.544      0.143      -1.815       0.290
x11           -0.2728      0.275     -0.990      0.338      -0.860       0.314
x12            0.3804      0.237      1.604      0.130      -0.125       0.886
x13           -0.1323      0.228     -0.580      0.571      -0.619       0.354
x14            0.3905      0.273      1.430      0.173      -0.191       0.972
x15            0.0511      0.228      0.224      0.826      -0.435       0.538
x16            0.2085      0.237      0.880      0.393      -0.296       0.713
x17           -0.4646      0.250     -1.859      0.083      -0.997       0.068
x18           -0.1080      0.223     -0.483      0.636      -0.584       0.368
x19           -0.0675      0.320     -0.211      0.836      -0.749       0.614
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     3.36e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.28e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [27]:
X_qsec_dummies = data_qsec_dummies.drop(columns=['qsec']).values
y_qsec_dummies = data_qsec_dummies['qsec'].values

In [28]:
X_qsec_dummies_train, X_qsec_dummies_test, y_qsec_dummies_train, y_qsec_dummies_test = train_test_split(X_qsec_dummies, y_qsec_dummies, train_size=0.4, random_state=137)
scaler = StandardScaler().fit(X_qsec_dummies_train)
X_qsec_dummies_train_scaled = scaler.transform(X_qsec_dummies_train)
X_qsec_dummies_test_scaled = scaler.transform(X_qsec_dummies_test)

In [29]:
X_qsec_dummies_train_sm = sm.add_constant(X_qsec_dummies_train_scaled)
modelo_qsec_dummies = sm.OLS(y_qsec_dummies_train, X_qsec_dummies_train_sm).fit()
data_qsec_dummies

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0   21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1   21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2   22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3   21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4   18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   
5   18.1  225.0  105  2.76  3.460  20.22   1   0  False   True  False    True   
6   14.3  360.0  245  3.21  3.570  15.84   0   0  False  False   True    True   
7   24.4  146.7   62  3.69  3.190  20.00   1   0   True  False  False   False   
8   22.8  140.8   95  3.92  3.150  22.90   1   0   True  False  False   False   
9   19.2  167.6  123  3.92  3.440  18.30   1   0  False   True  False   False   
10  17.8  167.6  123  3.92  3.440  18.90   1   0  False   True  False   False   
11  16.4  275.8  180  3.07  4.070  17.40   0   0  False  False   True    True   
12  17.3  275.8  180  3.07  3.730  17.60   0   0  False  False   True    True   
13  15.2  275.8  180  3.07  3.780  18.00   0   0  False  False   True    True   
14  10.4  472.0  205  2.93  5.250  17.98   0   0  False  False   True    True   
15  10.4  460.0  215  3.00  5.424  17.82   0   0  False  False   True    True   
16  14.7  440.0  230  3.23  5.345  17.42   0   0  False  False   True    True   
17  32.4   78.7   66  4.08  2.200  19.47   1   1   True  False  False   False   
18  30.4   75.7   52  4.93  1.615  18.52   1   1   True  False  False   False   
19  33.9   71.1   65  4.22  1.835  19.90   1   1   True  False  False   False   
20  21.5  120.1   97  3.70  2.465  20.01   1   0   True  False  False    True   
21  15.5  318.0  150  2.76  3.520  16.87   0   0  False  False   True    True   
22  15.2  304.0  150  3.15  3.435  17.30   0   0  False  False   True    True   
23  13.3  350.0  245  3.73  3.840  15.41   0   0  False  False   True    True   
24  19.2  400.0  175  3.08  3.845  17.05   0   0  False  False   True    True   
25  27.3   79.0   66  4.08  1.935  18.90   1   1   True  False  False   False   
26  26.0  120.3   91  4.43  2.140  16.70   0   1   True  False  False   False   
27  30.4   95.1  113  3.77  1.513  16.90   1   1   True  False  False   False   
28  15.8  351.0  264  4.22  3.170  14.50   0   1  False  False   True   False   
29  19.7  145.0  175  3.62  2.770  15.50   0   1  False   True  False   False   
30  15.0  301.0  335  3.54  3.570  14.60   0   1  False  False   True   False   
31  21.4  121.0  109  4.11  2.780  18.60   1   1   True  False  False   False   

    gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0     True   False   False   False   False    True   False   False  
1     True   False   False   False   False    True   False   False  
2     True   False    True   False   False   False   False   False  
3    False   False    True   False   False   False   False   False  
4    False   False   False    True   False   False   False   False  
5    False   False    True   False   False   False   False   False  
6    False   False   False   False   False    True   False   False  
7     True   False   False    True   False   False   False   False  
8     True   False   False    True   False   False   False   False  
9     True   False   False   False   False    True   False   False  
10    True   False   False   False   False    True   False   False  
11   False   False   False   False    True   False   False   False  
12   False   False   False   False    True   False   False   False  
13   False   False   False   False    True   False   False   False  
14   False   False   False   False   False    True   False   False  
15   False   False   False   False   False    True   False   False  
16   False   False   False   False   False    True   False   False  
17    True   False    True   False   False   False   False   False  
18    True   False  

In [30]:
modelo_qsec_dummies.summary()

/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/Users/paofigueroa/Documents/sem 5/Lab de aprendizaje estadístico/.venv/lib/python3.12/site-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = sel

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Wed, 17 Sep 2025   Prob (F-statistic):                nan
Time:                        19:35:02   Log-Likelihood:                 357.21
No. Observations:                  12   AIC:                            -690.4
Df Residuals:                       0   BIC:                            -684.6
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733        inf          0        nan         nan         nan
x1            -2.8275        inf         -0        nan         nan         nan
x2             2.1520        inf          0        nan         nan         nan
x3            -0.1677        inf         -0        nan         nan         nan
x4             3.7193        inf          0        nan         nan         nan
x5             0.0248        inf          0        nan         nan         nan
x6             0.0007        inf          0        nan         nan         nan
x7            -0.0380        inf         -0        nan         nan         nan
x8             0.7525        inf          0        nan         nan         nan
x9             0.7933        inf          0        nan         nan         nan
x10           -1.1698        inf         -0        nan         nan         nan
x11            1.1786        inf          0        nan         nan         nan
x12            0.4526        inf          0        nan         nan         nan
x13           -2.0851        inf         -0        nan         nan         nan
x14            1.5119        inf          0        nan         nan         nan
x15           -0.1165        inf         -0        nan         nan         nan
x16            0.5565        inf          0        nan         nan         nan
x17           -1.8342        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19            0.6457        inf          0        nan         nan         nan
==============================================================================
Omnibus:                        2.639   Durbin-Watson:                   2.045
Prob(Omnibus):                  0.267   Jarque-Bera (JB):                0.858
Skew:                           0.631   Prob(JB):                        0.651
Kurtosis:                       3.351   Cond. No.                         113.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

Es el mismo caso que con mpg. Hay demasiadas columnas, lo que hace que necesitemos de un modelo más grande. La realidad es que parece innecesario tener que usar dummies para variables que no son cualitativas, lo que hace que se necesiten más columnas cuando ni siquiera hay tantos datos. Es por eso que nos arojan un modelo perfecto con un overfit.

In [31]:
X_qsec_dummies_test_sm = sm.add_constant(X_qsec_dummies_test_scaled, has_constant='add')
y_qsec_dummies_pred = modelo_qsec_dummies.predict(X_qsec_dummies_test_sm)

In [32]:
r2_qsec_dummies_train = modelo_qsec_dummies.rsquared
r2_qsec_dummies_test = r2_score(y_qsec_dummies_test, y_qsec_dummies_pred)
print(f'R2 entrenamiento: {r2_qsec_dummies_train}')
print(f'R2 prueba: {r2_qsec_dummies_test}')

R2 entrenamiento: 1.0
R2 prueba: -2.860496844108422


Ejemplo perfecto de un ajuste muy bueno pero que hace overfitting, capturando más cosas de las necesarias, lo que hace que no se pueda ajustar bien a nuevos datos.

### 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

In [33]:
print("Sin regularización y sin dummies")
print(" r2 entrenamiento: ", r2_mpg_train)
print("r2 prueba", r2_mpg_test)
print("")
print("Con regularización")
print("r2 entrenamiento: ",resultados_mpg_train[0.1])
print("r2 prueba: ",resultados_mpg_test[0.1])
print("")
print("Sin regularización y dummies")
print("prueba: ", r2_mpg_dummies_train)
print("entrenamiento", r2_mpg_dummies_test)

Sin regularización y sin dummies
 r2 entrenamiento:  0.9900805114414537
r2 prueba -54.071785204187904

Con regularización
r2 entrenamiento:  0.9805234796986708
r2 prueba:  0.23411656981043205

Sin regularización y dummies
prueba:  1.0
entrenamiento 0.12691935910997298


Como hemos ido analizando, no porque sea un número perfecto significa que es el mejor. Aquí, evidentemente el usar las dummies nos es contraproducente, ya que está capturando datos irrelevantes en lugar de un patrón que generalice, creando un overfit que se ajusta demasiado a los datos de entrenamiento. Lo mejor sería usar el modelo sin dummies, ya que es casi perfecto y no captura tanto ruido. Ahora, en un modelo sin penalización, da una prueba muy mala, por lo que la mejor combinación sería un modelo que no use dummies pero que sí regularice

### 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

In [34]:
print("Sin regularización y sin dummies")
print(" r2 entrenamiento: ", r2_qsec_train)
print("r2 prueba", r2_qsec_test)
print("")
print("Con regularización")
print("r2 entrenamiento: ",resultados_qsec_train[0.1])
print("r2 prueba: ",resultados_qsec_test[0.1])
print("")
print("Sin regularización y dummies")
print("prueba: ", r2_qsec_dummies_train)
print("entrenamiento", r2_qsec_dummies_test)


Sin regularización y sin dummies
 r2 entrenamiento:  0.9787621925230521
r2 prueba 0.3761038962001224

Con regularización
r2 entrenamiento:  0.9773047022861615
r2 prueba:  0.5897331179193346

Sin regularización y dummies
prueba:  1.0
entrenamiento -2.860496844108422


Ocurre lo mismo que en el modelo mpg: hay un overfit en los datos con dummies porque hay más de lo que se necesita. Una vez más, lo mejor sería usar el modelo sin las dummies, el cual representa un 97%. Entre el modelo de mpg y qsec, guiándonos unicamente por el coeficiente de determinación y como, en general este define que tanto de la varianza de salida corresponde a la de entrada, sería mejor usar el modelo mpg. En conclusión de toda la actividad, los dummies per se no parecen ser el problema, sino el tamaño de la muestra: podríamos reducir variables o conseguir más datos y usar técnicas de regularización apropiadas. Las penalizaciones permiten forzar para quitarle peso a unas variables y tomar en cuenta otras, lo cual ayuda a la hora de ver el train contra test y, por último, al comparar los dos modelos , el  mpg muestra mejor estabilidad estadística y coeficientes más interpretables, sugiriendo que es más robusto para hacer predicciones confiables.